### PR TITLE
Add external generator with gap trigger for HF software triggers MC

### PR DESCRIPTION
**Pull request by fgrosa**
This PR:
- Adds an external generator for HF that allows us to trigger one event out of N (a la "gap-trigger" introduced by @ddobrigk in https://github.com/AliceO2Group/O2DPG/pull/933)
- Replaces the scripts for the MC productions for the HF software triggers according to the studies reported in https://alice.its.cern.ch/jira/browse/O2-2761
- Removes the script to run the MC productions for the HF software triggers with embedding with replacements since it is found not to work as expected at the moment (see https://alice.its.cern.ch/jira/browse/O2-3347)

### CODE REVIEW

**Review by gconesab — APPROVED** [review]

**Comment by chiarazampolli** [comment]
Hello @fgrosa , 
There are some errors in the CI. @benedikt-voelkel , could you please take a look?
Thanks!
Chiara

**Comment by fgrosa** [comment]
> Hello @fgrosa , There are some errors in the CI. @benedikt-voelkel , could you please take a look? Thanks! Chiara

Ciao @chiarazampolli, this is due to the new test functionality introduced by @benedikt-voelkel, I am on it

**Comment by fgrosa** [comment]
> Already discussed in WP4/14

@benedikt-voelkel I added the test macros (tested locally) in this [commit](https://github.com/AliceO2Group/O2DPG/pull/937/commits/4e4776da5b91ba3b6af43cc8660803b58dd561b8) you want to have a look

**Comment by benedikt-voelkel** [comment]
Thank you @fgrosa 

Do you know @TimoWilken  why the MC tests were not run? It should since there are changes in `MC/`, right?

**Comment by benedikt-voelkel** [comment]
@fgrosa 
It seems some conditions are not met for your tests.

Unfortunately it seems not all failed test log were printed. I will have a look

Another question: Should everything be tested for `External` **and** `Pythia8` or only for one of them?

**Comment by fgrosa** [comment]
Hi @benedikt-voelkel
> @fgrosa It seems some conditions are not met for your tests.
> 
> Unfortunately it seems not all failed test log were printed. I will have a look
Yes I saw (and fixed)
 
> Another question: Should everything be tested for `External` **and** `Pythia8` or only for one of them?

Indeed only `External` given the content of this PR (the `Pythia8` function is indeed empty now).

Cheers,
 Fabrizio

**Comment by benedikt-voelkel** [comment]
Ok, thanks.

Could you, in that case simply remove the `int Pythia8()` parts from the test macro?
In that way we know what was tested in hindsight.

**Comment by fgrosa** [comment]
> Ok, thanks.
> 
> Could you, in that case simply remove the `int Pythia8()` parts from the test macro? In that way we know what was tested in hindsight. Only one test is required at the moment.

done!

**Comment by fgrosa** [comment]
> > Ok, thanks.
> > Could you, in that case simply remove the `int Pythia8()` parts from the test macro? In that way we know what was tested in hindsight. Only one test is required at the moment.
> 
> done!

@benedikt-voelkel now the three tests related to this PR are passed, but still it's red. Can you please have a look that I did not miss anything? Thanks!

```
### Test PWG-related workflow creation ###

Test 1: /O2DPG/MC/run/PWGHF/run_pp_HFtriggers_Bforced_gaptrigger.sh[0;32m -> PASSED[0m
Test 2: /O2DPG/MC/run/PWGHF/run_pp_HFtriggers_bbbar_gaptrigger.sh[0;32m -> PASSED[0m
Test 3: /O2DPG/MC/run/PWGHF/run_pp_HFtriggers_ccbar_gaptrigger.sh[0;32m -> PASSED[0m


[0;32mAll required workflow tests successful[0m
```

**Comment by fgrosa** [comment]
> There is a problem with the`ccbar` related one I will take a look, for some reason the log is not printed. I will take a look and report back asap.
> 
> What you commented on is another test that makes sure that the sim workflows can be created. But for the `ini`s, one sees
> 
> ```shell
> Test 1: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_Bforced.ini with generator External�[0;32m -> PASSED�[0m
> Test 2: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_bbbar.ini with generator External�[0;32m -> PASSED�[0m
> Test 3: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini with generator External�[0;31m -> FAILED�[0m
> ```
> 
> Sorry, I will come back to you shortly!



> There is a problem with the`ccbar` related one I will take a look, for some reason the log is not printed. I will take a look and report back asap.
> 
> What you commented on is another test that makes sure that the sim workflows can be created. But for the `ini`s, one sees
> 
> ```shell
> Test 1: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_Bforced.ini with generator External�[0;32m -> PASSED�[0m
> Test 2: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_bbbar.ini with generator External�[0;32m -> PASSED�[0m
> Test 3: /O2DPG/MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini with generator External�[0;31m -> FAILED�[0m
> ```
> 
> Sorry, I will come back to you shortly!

You are right sorry, I only checked the bottom of the log! However, there is no message that should be there in each case of `return 1`, so indeed I am a bit puzzled.

File: MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini [hunk 1]
```diff
@@ -1,20 +1,7 @@
 ### The external generator derives from GeneratorPythia8.
-### This part configures the bits of the interface: configuration and user hooks
+[GeneratorExternal]
+fileName=${O2DPG_ROOT}/MC/config/PWGHF/pythia8/generator/generator_pythia8_gaptriggered_hf.C
```
Comment (benedikt-voelkel) [inline MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini:3]: That does not exist

Comment (benedikt-voelkel) [reply MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini:3]: That would have been in the `o2sim_serverlog`
I will update the test script to print that. So the error was acknowledged but not printed. Sorry

Comment (fgrosa) [reply MC/config/PWGHF/ini/GeneratorHFTrigger_ccbar.ini]: No problem and good catch! Fixed now :)

**Comment by benedikt-voelkel** [comment]
By the way, @fgrosa 
you can run the same thing locally by doing
```bash
O2DPG_TEST_REPO_DIR=${O2DPG_ROOT} O2DPG_TEST_HASH_BASE="HEAD~5" ${O2DPG_ROOT}/test/run_generator_tests.sh
```

I set `O2DPG_TEST_HASH_BASE="HEAD~5" but of course I don't know exactly how many commit are now between the `upstream@master` and your last developments

**Pull request merged**